Added thematic toggle and preferences page

Icyviolet23 opened the pull request:
Resolves #37 

Major changes

1. Added Preferences Model
2. Added one to one relation between the Preferences Model and the Abstract User model. Each User will have their own preferences.
3. Added Preferences ModelForm which will be used to collect user preferences
4. Added view to handle Preferences Form POST request, create an instance of the Preferences model and save it in the DB
5. Created HTML Template for preferences page
6. Edited HTML template for navigation.html to add navigation to preferences page
7. Added dark/light mode button and corresponding light/dark mode toggle functionality

Refer to the following DEMO:
https://www.youtube.com/watch?v=0qfaXa9EYyA&ab_channel=TzeHngLoke

Icyviolet23 committed:
Added model for preferences, tagged to to user mode, added view to fetch preferences and display them according. Added model form to allow users to customize their own preferences dynamically

kc12341 committed:
added toggle

kc12341 committed:
added click functionality to toggle, switching dark and light mode

kc12341 committed:
added dark/light mode toggle javascript functionality found in bootstrap documentation

kc12341 committed:
added dropdown with just text, but not integrated with javascript toggle functionality yet

kc12341 committed:
added active mode status label on the dropdown, modified JS to account for text labels rather than svgs

Icyviolet23 committed:
integrated Kyle's changes by toggling light and dark mode based on user preferences. The mode will be automatically set to the correpsonding preferred mode

brylie reviewed (changes requested):
Overall, this is looking great! I've just added a few comments about internationalization tags so our app can be localized in different languages.

https://docs.djangoproject.com/en/5.0/topics/i18n/translation/#translation

kc12341 committed:
replaced bootstrap implementation with self implementation for more flexibility

kc12341 committed:
removed auto from dropdown since it was not necessary

kc12341 committed:
saves theme to local storage and reduced code redundancy in toggleTheme function

kc12341 committed:
removed console logging statements

kc12341 committed:
Setting color mode directly without JS

Changing body attribute data-bs-theme to directly change rather than using JS

Co-authored-by: Brylie Christopher Oxley <brylie@protonmail.com>

kc12341 committed:
Removing script from base.html

Directly changing color mode by accessing value in template makes script unnecessary

Co-authored-by: Brylie Christopher Oxley <brylie@protonmail.com>

kc12341 committed:
accepted suggestions to remove script from base.html

kc12341 committed:
added suggested changes to remove script from base.html from code repo owner

kc12341 committed:
Enumeration changes for language and color mode changes

Using TextChoices enumeration for choices rather than lists

Co-authored-by: Brylie Christopher Oxley <brylie@protonmail.com>

kc12341 committed:
new tests for added toggle functionality

kc12341 committed:
Merge branch 'theme-toggle' of https://github.com/adrienneli104/caregiving into theme-toggle

brylie commented on preferences/views.py:
```diff
@@ -0,0 +1,43 @@
+        # null check for preferences field
+        if request.user.preferences:
+            fields = [(field.name, field.value_to_string(request.user.preferences)) for field in Preferences._meta.fields]
+            context['fields'] = fields
```
I think we only need to return a form instance pre-populated with the existing user preferences.Django takes care of everything else.
```suggestion
            context['form'] = PreferencesForm(instance=request.user.preferences)
```

brylie commented on preferences/views.py:
```diff
@@ -0,0 +1,43 @@
+        context['form'] = PreferencesForm ()
+        # null check for preferences field
+        if request.user.preferences:
+            fields = [(field.name, field.value_to_string(request.user.preferences)) for field in Preferences._meta.fields]
```
This is probably unnecessary 
```suggestion
```

brylie commented on preferences/forms.py:
```diff
@@ -0,0 +1,14 @@
+class PreferencesForm (ModelForm):
+    class Meta:
+        model = Preferences
+        fields = ('Language', 'Mode')
```
These fields should match the model fields
```suggestion
        fields = ('language', 'color_mode')
```

brylie commented on preferences/models.py:
```diff
@@ -0,0 +1,17 @@
+from django.conf import settings
+
+# for user preferences
+class Preferences (models.Model):
```
I believe adding a ‘user = models.OneToOne(…)’ field on the ‘Preferences’ model will make the view.py code simpler.

brylie commented on accounts/models.py:
```diff
@@ -10,3 +11,8 @@ class Meta:
         return self.username
+
+    #store user preferences in a one to one mapping
+    preferences = models.OneToOneField(Preferences, on_delete=models.CASCADE, blank=True, null=True) 
```
Let’s define the recon the Preferences model rather than the user. Then, use a ‘reverse_name=“preferences”’ on the ‘user’ OneToOne field attached to the Preferences model. That way, we assign the user to the Preferences instance directly in view.py.

brylie replied:
```suggestion
```

brylie commented on preferences/views.py:
```diff
@@ -0,0 +1,43 @@
+                Mode = form.cleaned_data['Mode'],
+                )
+
+    preferences.save ()
```
This can just be ‘form.save()’, so there is no need to create a temporary Preferences instance above
```suggestion
    preferences = form.save()
```

brylie reviewed (changes requested):
It looks like there are some conflicts in `settings.py`. Let's resolve those conflicts so we can merge this PR soon. Overall, the PR is looking good.

brylie commented:
Closing since this PR had unresolved conflicts
